compositor_pipeline/src/audio_mixer/internal_audio_mixer.rs

```diff
@@ -1,61 +1,177 @@
-use std::{collections::HashMap, time::Duration};
+use std::{
+    cmp::{max, min},
+    collections::HashMap,
+    sync::Arc,
+};
 
 use compositor_render::{error::UpdateSceneError, OutputId};
 
-use super::types::{AudioMixingParams, AudioSamplesSet, OutputSamples};
+use crate::audio_mixer::types::{AudioSamples, AudioSamplesBatch};
+
+use super::types::{AudioChannels, AudioMixingParams, AudioSamplesSet, OutputSamples};
 
 #[derive(Debug)]
 struct OutputInfo {
     audio: AudioMixingParams,
-    last_batch_pts: Option<Duration>,
+    channels: AudioChannels,
 }
 
 #[derive(Debug)]
 pub(super) struct InternalAudioMixer {
     outputs: HashMap<OutputId, OutputInfo>,
+    output_sample_rate: u32,
 }
 
 impl InternalAudioMixer {
-    pub fn new() -> Self {
+    pub fn new(output_sample_rate: u32) -> Self {
         Self {
             outputs: HashMap::new(),
+            output_sample_rate,
         }
     }
 
     pub fn mix_samples(&mut self, samples_set: AudioSamplesSet) -> OutputSamples {
-        let get_batch = |output_info: &OutputInfo| {
-            let input_id = &output_info.audio.inputs.first()?.input_id;
-            samples_set.samples.get(input_id)?.iter().find(|batch| {
-                let last_pts = match output_info.last_batch_pts {
-                    Some(pts) => pts,
-                    None => return true,
+        trait Sample {
+            fn add_assign(&mut self, rhs: Self);
+            fn div(self, other: i32) -> Self;
+            fn mul(self, other: f32) -> Self;
+        }
+
+        fn mix_output<T: Clone + Copy + Default + Sample, F: Fn(&AudioSamplesBatch) -> Vec<T>>(
+            output_info: &OutputInfo,
+            output_sample_rate: u32,
+            samples_set: &AudioSamplesSet,
+            get_samples: F,
+        ) -> Vec<T> {
+            let samples_count = samples_set
+                .end_pts
+                .saturating_sub(samples_set.start_pts)
+                .as_secs_f64()
+                * output_sample_rate as f64;
+            let mut mixed = vec![T::default(); samples_count as usize];
+            for input in &output_info.audio.inputs {
+                let Some(input_batches) = samples_set.samples.get(&input.input_id) else {
+                    continue;
                 };
+                input_batches.iter().for_each(|batch| {
+                    let batch_samples = get_samples(batch);
+                    let start_sample = (batch
```

Comment:
incorrect name, this index in a batch that should represent the first sample in the output

also, improve formatting here